### PR TITLE
Feature/evidence

### PR DESCRIPTION
**Pull request by matthewghgriffiths**
Pull request for some changes to `graphical`

changes:

- changed API for calling factors now call signature is
  `factor(values: Dict[Variable, np.ndarray], axis: Union[False, None, int, Tuple[int, ....]) -> FactorValue:`
   `axis` controls the dimensionality of the output, so `factor(values, axis) == np.sum(factor(values, axis=False), axis=axis)`
   if `axis` is false the full dimensionality of the output is returned.
- FactorValue is now a subclass of a `numpy.ndarray`, so it functions as a normal number/array, except it has an additional attribute `deterministic_values` which is a dictionary that stores any deterministic values
- Added an API for passing analytic gradients and jacobians
  - Created `FactorJacobian` and `DeterministicFactorJacobian` classes for these objects
  - call signature is, 
    ```
    factor.func_jacobian(
            values: Dict[Variable, np.ndarray], 
            variables: Tuple[Variable], 
            axis: Union[False, None, int, Tuple[int, ....],
            _calc_deterministic: bool = True, 
    ) -> Tuple[FactorValue, Dict[Variable, FactorValue]]:
    ```
    where variables controls which passed variables to calculate the gradients/Jacobians for. _calc_deterministic determines whether the Jacobians of the deterministic variables are calculated and returned.
- added gradient methods `logpdf_gradient` for NormalMessage and GammaMessage
  - Also added `as_factor` method so that messages can be used as factors
- Created `MeanField` class for encoding mean-field approximations to factors
- Created `EPOptimiser` class for performing EP inference
  - this method as a `log_evidence` property for returning the estimated log evidence
- Created `EPHistory` class for saving EP history and also performing convergence checks
  - added `kl_tol` and `evidence_tol` as keywords as controls for convergence checks
  - `kl_tol` works best as a convergence check (at least for Laplace's approximation)
- fixed a variety of bugs and refactored code

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,135 +1,139 @@
...
+# docs
+docs/_build
+docs/_static
+docs/_templates
+docs/generated
+docs/api/generated
+test_autofit/unit/samples.csv
+test_autofit/unit/non_linear/files/pickles/test.pickle
+test_autofit/unit/non_linear/nest/files/dynesty/output/dynesty_static[nlive_150__bound_multi_vol_dec_0.5_vol_check_2.0__enlarge_1.0__sample_auto]/samples/dynesty.pickle
+>>>>>>> develop
```
Comment (rhayes777) [inline .gitignore:139]: There's a stray conflict annotation here

File: autofit/graphical/__init__.py [hunk 2]
```diff
@@ -1,6 +1,13 @@
-from .declarative import ModelFactor, ModelFactorCollection
-from .factor_graphs import Factor, FactorGraph, AbstractFactor, FactorValue
-from .mean_field import FactorApproximation, MeanFieldApproximation
-from .messages import FixedMessage, NormalMessage, GammaMessage, AbstractMessage
-from .optimise import OptFactor, lstsq_laplace_factor_approx
-from .sampling import ImportanceSampler, project_factor_approx_sample
+from .declarative import ModelFactor, ModelFactorCollection
+from .factor_graphs import \
```
Comment (rhayes777) [inline autofit/graphical/__init__.py:2]: If you're importing lots of stuff from a single module it's sometimes nice to alias the module and use that.
e.g.
```python
from . import factor_graphs as fg

fg.Factor

**Review by rhayes777 — COMMENTED** [review]
This should actually be into develop. I can see a bunch of my changes in here too - I guess @Jammy2211 hasn't done a release on AutoFit recently?

**Comment by Jammy2211** [comment]
> This should actually be into develop. I can see a bunch of my changes in here too - I guess @Jammy2211 hasn't done a release on AutoFit recently?

I did one yesterday, the sensitivity stuff is in there so pretty up to date?

File: .gitignore (hunk 1 above)
Comment (matthewghgriffiths) [reply .gitignore:139]: Yeah that shouldn't be there. What's the process for removing it from the pull request?

**Review by rhayes777 — APPROVED** [review]
Will add some more comments in a bit got sidetracked

**Comment by rhayes777** [comment]
If you push to the branch it will update the PR

File: autofit/graphical/__init__.py (hunk 2 above)
Comment (matthewghgriffiths) [reply autofit/graphical/__init__.py:2]: yeah I can see that, but in this case it's useful to directly expose those imports so only need to import `autofit.graphical`

```
import autofit.graphical as graph

graph.Factor
```

File: autofit/graphical/expectation_propagation.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    def variables(self): 
+        return self.factor_graph.variables
+
+    @property
+    def deterministic_variables(self): 
+        return self.factor_graph.deterministic_variables
+
+    @property
+    def variable_names(self) -> Dict[str, Variable]: 
+        return self.factor_graph.variable_names
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:53]: I agree that this attribute name is probably not ideal. 

what about `name_variable` or `name_variable_dict` (I prefer `name_variable`)?

File: autofit/graphical/expectation_propagation.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+        return cls(
+            factor_graph,
+            factor_mean_field)
+
+    from_kws = from_approx_dists
+    
+    def factor_approximation(self, factor: Factor) -> FactorApproximation:
+        factor_mean_field = self.factor_mean_field
+        factor_dist = factor_mean_field.pop(factor)
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:82]: Yeah this is a gotcha. 

`self.factor_mean_field` is a property that returns a copy of the `self._factor_mean_field` dictionary

I suppose to make it clearer it should at least by commented on, or alternatively,
`factor_mean_field = self.factor_mean_field.copy()` -> means dictionary is copied unnecessarily
`factor_mean_field = self._factor_mean_field.copy()` -> relies on private implementation details

File: autofit/graphical/expectation_propagation.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    def variable_evidence(self) -> Dict[Variable, np.ndarray]:
+        return {
+            v: AbstractMessage.log_normalisation(*ms)
+            for v, ms in self.variable_messages.items()}
+
+    @property 
+    def factor_evidence(self) -> Dict[Factor, np.ndarray]:
+        return {
+            factor: meanfield.log_norm 
+            for factor, meanfield in self.factor_mean_field.items()}
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:144]: Our biggest fights are definitely going to be around multi-line bracket-styling :P  

There is a neatness to put the closing brace on a new line, particularly for highly nested bracketing, however I find that splitting up code over too many lines can also reduce readability, by reducing the amount of logic that can be viewed simultaneously and breaking up expressions into too small chunks that lose context.

Is there an explicit Autofit style guide?

File: autofit/graphical/expectation_propagation.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+                ∫ ∏_{j ∈ a} m_{i → a} (xᵢ) fₐ(xₐ) dxₐ
+        Zₐ = -----------------------------------------
+                             ∏_{j ∈ a} Zⱼ
+
+        Evidence for model
+
+        Z = ∏ᵢ Zᵢ ∏ₐ Zₐ
+        """
+        variable_evidence = {
+            v: np.sum(Zi) for v, Zi in self.variable_evidence.items()}
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:166]: In this case Zi is representing the variable evidence, and this quantity is canonically notationally represented by a capital Z. 

In this case it's not massively important as the context makes it pretty clear. 

also [relevant](https://twitter.com/tesseralis/status/1259306924312391686)

File: autofit/graphical/expectation_propagation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py:53]: I prefer name_variable_dict. To me name_variable sounds like a kind of variable

File: autofit/graphical/expectation_propagation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py:82]: Did you need to use the pop function? If you don't use it there's no side effect?

It's fine to use the private version if it's internal usage

File: autofit/graphical/expectation_propagation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py:144]: Studies show longer lines in some text make something quicker to read but harder to understand; in software I think understanding is key.

I guess the advantage of these broken apart statements is that you can tell at a glance what's going on. Moreover, you're likely to only care about the function you're looking and perhaps some caller which is not necessarily near that function. Provided the function easily fits on the screen breaking it up like this isn't an issue.

There is no style guide but I may well publish one.

File: autofit/graphical/expectation_propagation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py:166]: Perhaps in mathematical notation but this is Python!

File: autofit/graphical/expectation_propagation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:82]: Yes the pop function is needed.

File: autofit/graphical/__init__.py (hunk 2 above)
Comment (rhayes777) [reply autofit/graphical/__init__.py:2]: Oh yeah you're totally right I didn't pay attention to the filename!

File: autofit/graphical/mean_field.py [hunk 7]
```diff
@@ -1,291 +1,650 @@
...
+
+            if not q_f1.is_valid:
+                success = False
+                messages += (
+                    f"factor projection for {v} with {factor_approx.factor} is invalid",)
+
+            factor_projection[v] = q_f1
+
+            #             if isinstance(q_cavity, AbstractMessage):
+#             log_norm += (np.sum(q_fit.log_norm)
```
Comment (rhayes777) [inline autofit/graphical/mean_field.py:72]: Could you remove commented code?

File: autofit/graphical/mean_field.py [hunk 8]
```diff
@@ -1,291 +1,650 @@
...
+        factor_dist=MeanField(factor_projection),
+        model_dist=MeanField(model_dist),
+#         log_norm=log_norm
+    )
+    status = Status(success, messages)
+
+    return projection, status
+
+    
+class MeanField(Dict[Variable, AbstractMessage], Factor):
```
Comment (rhayes777) [inline autofit/graphical/mean_field.py:97]: You can inherit from a type annotation? Nuts.

File: autofit/graphical/mean_field.py [hunk 9]
```diff
@@ -1,291 +1,650 @@
...
+    project = project_mean_field
+
+    def __repr__(self):
+        # TODO make this nicer
+        return f"{type(self).__name__}({self.factor}, ...)"
+
+    
+class MeanFieldApproximation:
+    '''
+    TODO: delete this class
```
Comment (rhayes777) [inline autofit/graphical/mean_field.py:503]: Delete this class?

File: autofit/graphical/messages/abstract.py [hunk 10]
```diff
@@ -1,256 +1,389 @@
...
+from typing import Optional, Tuple, Union
+from inspect import getfullargspec
+
+import numpy as np
+
+
+class AbstractMessage(ABC):
+    log_base_measure: float
+    _multivariate: bool = False
+    _parameter_support: Optional[Tuple[Tuple[float, float], ...]] = None
```
Comment (rhayes777) [inline autofit/graphical/messages/abstract.py:14]: lol those are some pretty complicated types

File: autofit/graphical/messages/abstract.py [hunk 11]
```diff
@@ -1,256 +1,389 @@
...
+            else:
+                logl0 = self.logpdf(x0)
+                l0 = logl0.sum()
+                grad_logl = np.empty_like(x0)
+                hess_logl = np.empty_like(x0)
+                with np.nditer(x0, flags=['multi_index'], op_flags=['readwrite']) as it:
+                    for xv in it:
+                        xv += eps
+                        l1 = self.logpdf(x0).sum()
+                        xv -= 2 * eps
```
Comment (rhayes777) [inline autofit/graphical/messages/abstract.py:237]: Looks like l1 == l2?

File: autofit/graphical/messages/abstract.py [hunk 12]
```diff
@@ -1,256 +1,389 @@
...
+        tx = cls.to_canonical_form(samples)
+        w /= norm
+        suff_stats = (tx * w[None, ...]).mean(1)
+
+        assert np.isfinite(suff_stats).all()
+        return cls.from_sufficient_statistics(suff_stats, log_norm=log_norm)
+
+    @classmethod
+    def from_mode(cls, mode, covariance):
+        pass
```
Comment (rhayes777) [inline autofit/graphical/messages/abstract.py:282]: Might be sensible to raise NotImplemented()

File: autofit/graphical/messages/abstract.py [hunk 13]
```diff
@@ -1,256 +1,389 @@
...
+                 for p, support in zip(self.parameters, self._parameter_support)))
+        elif self.ndim:
+            return np.array(True, dtype=bool, ndmin=self.ndim)
+        return np.array([True])
+
+    def check_finite(self) -> np.ndarray:
+        return np.isfinite(self.natural_parameters).all(0)
+
+    def check_valid(self):
+        return self.check_finite() & self.check_support()
```
Comment (rhayes777) [inline autofit/graphical/messages/abstract.py:342]: Did you mean to use bitwise operators here and 332? Are the objects implementing some custom boolean logic (e.g. numpy)?

File: autofit/graphical/messages/fixed.py [hunk 14]
```diff
@@ -1,72 +1,78 @@
...
+    @classmethod
+    def invert_sufficient_statistics(cls, suff_stats):
+        return suff_stats
+
+    def sample(self, n_samples=None):
+        """
+        Rely on array broadcasting to get fixed values to
+        calculate correctly
+        """
+        if n_samples is None:
```
Comment (rhayes777) [inline autofit/graphical/messages/fixed.py:41]: Seems odd that n_samples is only used to decide whether to index. It's effectively a boolean but sounds like an integer

File: test_autofit/unit/graphical/regression/test_linear_regression.py [hunk 15]
```diff
@@ -1,86 +1,219 @@
...
+
+def test_jacobian(
+        a_,
+        b_,
+        x_,
+        z_,
+        linear_factor,
+        linear_factor_jac,
+):
+    n, m, d = 5, 4, 3
```
Comment (rhayes777) [inline test_autofit/unit/graphical/regression/test_linear_regression.py:84]: I know the tests I've written for the graphical stuff are pretty big, but you should try to make each group of assertions into one test. You can set up and common state you'll use across multiple tests in fixtures and then make each test just a few lines. That way, when tests fail, they give you more information about what's going wrong.

File: test_autofit/unit/graphical/regression/test_linear_regression.py [hunk 16]
```diff
@@ -1,86 +1,219 @@
...
+
+    # history = {}
+    # n_iter = 3
+
+    # for i in range(n_iter):
+    #     for factor in model.factors:
+    #         # We have reduced the entire EP step into a single function
+    #         model_approx, _ = mp.sampling.project_model(
+    #             model_approx,
+    #             factor,
```
Comment (rhayes777) [inline test_autofit/unit/graphical/regression/test_linear_regression.py:202]: How come this is commented out?

File: test_autofit/unit/graphical/regression/test_logistic_regression.py [hunk 17]
```diff
@@ -1,85 +1,153 @@
...
+
+    q_z = model_approx.mean_field[z_]
+    y = model_approx.mean_field[y_].mean
+    y_pred = q_z.mean > 0
+    (tpr, fpr), (fnr, tnr) = np.dot(
+        np.array([y, 1-y]).reshape(2, -1),
+        np.array([y_pred, 1-y_pred]).reshape(2, -1).T)
+
+    accuracy = (tpr + tnr) / (tpr + fpr + fnr + tnr)
+    assert 0.9 > accuracy > 0.7
```
Comment (rhayes777) [inline test_autofit/unit/graphical/regression/test_logistic_regression.py:147]: Sounds like a more robust metric - remove my crappy assertions if you can replace them with something less arbitrary

File: test_autofit/unit/graphical/test_transform.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+    
+    b = np.random.rand(d, d + 1)
+    assert np.allclose(cho_factor * b, U @ b)
+    assert np.allclose(cho_factor.ldiv(b), iU @ b)
+
+    b = np.random.rand(d + 1, d)
+    assert np.allclose(b * cho_factor, b @ U)
+    assert np.allclose(b / cho_factor, b @ iU)
+
+def test_DiagonalTransform():
```
Comment (rhayes777) [inline test_autofit/unit/graphical/test_transform.py:40]: function names should be lower case.

You can group tests into test cases which are classes. In this case you could then break this test into several test methods within the test case

**Review by rhayes777 — APPROVED** [review]
I'll aim to take a deep dive next week

File: autofit/graphical/expectation_propagation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jammy2211) [reply autofit/graphical/expectation_propagation.py]: Why not just use black to auto format everything? :P

[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py]: Worst thing I ever suggested!

[the remark this answers is not in the document]
Comment (Jammy2211) [reply autofit/graphical/expectation_propagation.py]: I'd back Rich on this one btw, the bracket spacing took me a while to get used to but now I did I couldn't go back.

File: autofit/graphical/messages/abstract.py (hunk 11 above)
Comment (matthewghgriffiths) [reply autofit/graphical/messages/abstract.py:237]: `xv` is a single element of the array `x0` so it's been changed by +/-  `eps`

File: autofit/graphical/messages/abstract.py (hunk 13 above)
Comment (matthewghgriffiths) [reply autofit/graphical/messages/abstract.py:342]: yes, there's numpy boolean logic going on here

File: autofit/graphical/expectation_propagation.py [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    @abstractmethod
+    def optimise(
+            self, 
+            factor: Factor, 
+            model_approx: EPMeanField, 
+            status: Optional[Status] = None
+    ) -> Tuple[EPMeanField, Status]:
+        pass
+
+EPCallBack = Callable[[Factor, EPMeanField, Status], bool]
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/expectation_propagation.py:240]: It's an alias for a 'type'

File: autofit/graphical/mean_field.py (hunk 9 above)
Comment (matthewghgriffiths) [reply autofit/graphical/mean_field.py:503]: Yeah I need to update the whole test suite to do that...

File: autofit/graphical/expectation_propagation.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rhayes777) [reply autofit/graphical/expectation_propagation.py:240]: Ah yeah I guess that makes sense

File: test_autofit/unit/graphical/regression/test_linear_regression.py (hunk 16 above)
Comment (matthewghgriffiths) [reply test_autofit/unit/graphical/regression/test_linear_regression.py:202]: that functionality is superseded `ep_opt.run` method

File: autofit/graphical/mean_field.py [hunk 20]
```diff
@@ -1,291 +1,650 @@
...
+
+        super().__init__(**factor._kwargs)
+
+    @property
+    def variables(self):
+        return self.factor.variables
+
+    @property
+    def deterministic_variables(self):
+        return self.factor.deterministic_variables
```
[the remark this answers is not in the document]
Comment (matthewghgriffiths) [reply autofit/graphical/mean_field.py:371]: Yeah, I don't like the exceptions that it generates, I feel explicitly defining the properties is clearer for understanding and debugging.

File: autofit/graphical/messages/fixed.py (hunk 14 above)
Comment (matthewghgriffiths) [reply autofit/graphical/messages/fixed.py:41]: I kinda know what you mean, the syntax is supposed enable easy sampling of single values of the message (`n_samples=None`) and multiple samples (`n_samples=n:int`) i.e

`np.array([message.sample(n_samples=None) for _ in range(n_samples)])`

**Pull request merged**